File: design-pattern-go/src/test/patterns/behavioral/chain_responsibility/chain_responsibility_test.go
package chainresponsibility

import (
	chainresponsibility "design_patterns_training/patterns/behavioral/chain_responsibility"
	"testing"
)

func TestShouldHandlerBeCalledWhenHandleIsCalled(
	t *testing.T) {
	var wasCalled = false
	var child = chainresponsibility.NewLeaf(func() {
		wasCalled = true
	})
	child.Handle()
	if !wasCalled {
		t.Fatal("Handler was not called")
	}
}
func TestShouldCallParentHandlerWhenNoHandleIsSetButCalled(
	t *testing.T) {
	var wasCalled = false
	var child = chainresponsibility.NewLeaf(nil)
	var parent = chainresponsibility.NewContainer(func() {
		wasCalled = true
	})
	parent.Add(&child)
	child.Handle()
	if !wasCalled {
		t.Fatal("Handler was not called")
	}
}
func TestShouldCallTheHigherParentHandlerMethodWhenNoOneIsSet(
	t *testing.T) {
	var wasCalled = false
	var lowerChild = chainresponsibility.NewLeaf(nil)
	var lowerParent = chainresponsibility.NewContainer(nil)
	var higherParent = chainresponsibility.NewContainer(func() {
		wasCalled = true
	})
	higherParent.Add(&lowerParent.Leaf)
	lowerParent.Add(&lowerChild)
	lowerChild.Handle()
	if !wasCalled {
		t.Fatal("Handle was not called")
	}
}

File: design-pattern-jvm/app/src/test/java/design/pattern/kotlin/patterns/behavioral/chain_responsibility/ChainResponsibilityTest.kt
package design.pattern.kotlin.patterns.behavioral.chain_responsibility

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ChainResponsibilityTest {
    internal inner class ValueWrapper(private var wasCalled: Boolean) {
        fun setValue(value: Boolean) {
            wasCalled = value
        }

        fun hasBeenCalled(): Boolean {
            return wasCalled
        }
    }

    @Test
    fun shouldHandlerBeCalledWhenHandleIsCalled() {
        val wasCalled: ValueWrapper = ValueWrapper(false)
        val child = Leaf {
            wasCalled.setValue(true)
            true
        }
        child.handle()
        Assertions.assertTrue(wasCalled.hasBeenCalled())
    }

    @Test
    fun shouldCallParentHandlerWhenNoHandleIsSetButCalled() {
        val wasCalled: ValueWrapper = ValueWrapper(false)
        val child = Leaf(null)
        val parent = Container {
            wasCalled.setValue(true)
            true
        }
        parent.add(child)
        child.handle()
        Assertions.assertTrue(wasCalled.hasBeenCalled())
    }

    @Test
    fun shouldCallTheHigherParentHandlerMethodWhenNoneIsSer() {
        val wasCalled: ValueWrapper = ValueWrapper(false)
        val child = Leaf(null)
        val lowerParent = Container(null)
        val higherParent = Container {
            wasCalled.setValue(true)
            true
        }
        higherParent.add(lowerParent)
        lowerParent.add(child)
        child.handle()
        Assertions.assertTrue(wasCalled.hasBeenCalled())
    }
}